### PR TITLE
Stackoverflow fix

### LINKED ISSUE #1704: thread '<main>' panicked at 'called when calling 'trace_transaction'
Attempting to call trace_transaction causes an exception in the latest Parity/v1.3.0-unstable-a76981a-20160724/x86_64 as below:

C:\Users\MT4\.cargo\bin>parity --tracing on
2016-07-24 09:54:18 Starting Parity/v1.3.0-unstable-a76981a-20160724/x86_64-windows-msvc/rustc1.10.0
2016-07-24 09:54:18 Using state DB journalling strategy fast
2016-07-24 09:54:18 Configured for Frontier/Homestead using Ethash engine
2016-07-24 09:54:18 Public node URL: enode://7c81c7305c075e81399897e500099565e733cfb67e7c2034c7b52fc73b769bd832f6c29d31c07541c57d083b8d1f01228a01257fa794c87bea36a7c2bf99a771@127.0.0.1:30303
2016-07-24 09:54:49     0/ 3/25 peers     37 MiB db   10 KiB chain    2 KiB queue   10 KiB sync
2016-07-24 09:55:19     0/ 3/25 peers     37 MiB db   10 KiB chain    2 KiB queue   10 KiB sync
2016-07-24 09:55:50     0/ 3/25 peers     37 MiB db   10 KiB chain    2 KiB queue   10 KiB sync
2016-07-24 09:56:20     0/ 3/25 peers     37 MiB db   12 KiB chain    2 KiB queue   10 KiB sync
thread '<unnamed>' panicked at 'DecodeError: RlpExpectedToBeList', Users\MT4\.cargo\git\checkouts\parity-6ba620efeffb2216\master\util\src\rlp\rlpin.rs:119
note: Run with `RUST_BACKTRACE=1` for a backtrace.
2016-07-24 09:56:29 Finishing work, please wait...
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Closed', ../src/libcore\result.rs:785

### PR DESCRIPTION
**Pull request by debris**
fixes #1704

### CODE REVIEW

File: ethcore/src/state.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub struct ApplyOutcome {
 	/// The receipt for the applied transaction.
 	pub receipt: Receipt,
 	/// The trace for the applied transaction, if None if tracing is disabled.
```
Comment (tomusdrw) [inline ethcore/src/state.rs]: Comment should be updated.

**Comment by gavofyork** [comment]
this alters the user-visible JSONRPC format; can we do a follow-up PR prior to 1.3 to reinstate the original trace format?

**Comment by debris** [comment]
Only `trace_call`, `trace_rawTransaction` and trace_replayTransaciton` are changed and there is no way of preserving the old format cause serialization of data will overflow the stack.

`rawTransaction` and `replayTransaction` were merged 2 days ago, so they were not yet in release version. So the only broken method is `trace_call`, but afaik it's not used by exchanges/block explorers.

**Comment by gavofyork** [comment]
it is used by some of my ui, and in general the heirarchical JSON more closely represents the underlying JSON. i'd prefer not to cut any corners on standards just because serde's model is inadequate. ultimately, these should be adopted across clients and it's hard to make a case for a substandard API/format on the argument of "a library of one of the implementation makes this format easier".

**Comment by debris** [comment]
> it is used by some of my ui, and in general the heirarchical JSON more closely represents the underlying information

I agree, but this model does not work now. It overflows very often when serializing to rlp / serializing to json

> i'd prefer not to cut any corners on standards just because serde's model is inadequate.

We already have one format of returning [traces](https://gist.github.com/debris/051c131e877affefeab4553509640a43) on which we agreed on few months ago. This format is already used be exchanges / block explorers. I don't see a reason to have another one

**Comment by gavofyork** [comment]
they fulfil two different needs though. the format to which you linked must handle the case of returning a collection of individual messages/creates/suicides from across many transactions.

when tracing a transaction, that is not needed; rather a hierarchical format makes more sense since it directly reflects the structure of the call graph.

**Comment by arkpar** [comment]
I'd prefer to have a single output format for all types of queries. We might also want to filter individual tx traces by address in the future for example.

**Comment by gavofyork** [comment]
we have different outputs for different RPC calls depending on what data they're providing. this is no different.

**Pull request merged**